(BOLT-1472) Setup Kerberos in Container Tests (#999)
puppetlabs/bolt: developer-docs/kerberos.md

- `bolt command run 'whoami' --targets winrm://dc.domain.com` to connect over HTTPS (`--no-ssl-verify` may be required if the target uses a self-signed certificate)
- `bolt command run 'whoami' --targets winrm://dc.domain.com --no-ssl` to connect over HTTP

### Testing with Docker

> In Bolt testing atop Docker containers, [Samba server](https://www.samba.org/) is setup on Linux to approximate an Active Directory setup, which also includes DNS and LDAP support. Given Kerberos has very strict requirements around computer identity (via DNS), Docker user-defined networks with custom DNS and subnets are easier to setup than running in an arbitrary network environment. This lends itself well to an automated and reproducable ephemerable Kerberos environment.

Reviewer: ```suggestion
In Bolt testing atop Docker containers, [Samba server](https://www.samba.org/) is setup on Linux to approximate an Active Directory setup, which also includes DNS and LDAP support. Given Kerberos has very strict requirements around computer identity (via DNS), Docker user-defined networks with custom DNS and subnets are easier to setup than running in an arbitrary network environment. This lends itself well to an automated and reproducable ephemerable Kerberos environment.
```
```suggestion
In Bolt testing atop Docker containers, [Samba server](https://www.samba.org/) is setup on Linux to approximate an Active Directory setup, which also includes DNS and LDAP support. Given Kerberos has very strict requirements around computer identity (via DNS), Docker user-defined networks with custom DNS and subnets are easier to setup than running in an arbitrary network environment. This lends itself well to an automated and reproducable ephemeral Kerberos environment.
```

###### External Ports

* 45985 (tcp) - WinRM HTTP (internally 5985)
* 45986 (tcp) - WinRM HTTPS (internally 5986)

###### Interactive Shell Access

To access the shell, use `/bin/bash` like:

> docker-compose -f spec/docker-compose.yml exec omiserver /bin/bash

Useful tooling on the instance includes:

* [`host`](https://linux.die.net/man/1/host) - DNS lookup utility
* [`klist`](https://web.mit.edu/kerberos/krb5-devel/doc/user/user_commands/klist.html) - check Kerberos tickets
* [`realm`](https://www.systutorials.com/docs/linux/man/8-realm/) - manages enrollment in Kerberos realms and Active Directory domains
* [`net`](https://www.samba.org/samba/docs/current/man-html/net.8.html) - designed to work like the `net` tool on Windows
* [`pwsh`](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell-core-on-linux?view=powershell-6) - PowerShell 6
* [`omicli`](https://github.com/microsoft/omi/blob/master/Unix/cli/examples.txt) - client tool used to verify basic OMI server functionality

#### TravisCI automated testing

At present, Travis setup will:

> * Ensure that it can refer to itself as `samba-ad.bolt.test` and `omiserver.bolt.test` (the same name that the containesr refer to themselves inside the Docker UDN)

Reviewer: ```suggestion
* Ensure that it can refer to itself as `samba-ad.bolt.test` and `omiserver.bolt.test` (the same name that the containers refer to themselves inside the Docker UDN)
```

* Install the Kerberos client package
* Configure the Kerberos client with the approriate server (`samba-ad.bolt.test`) for the realm `BOLT.TEST`

> Automated tests (in `spec/bolt/transport/winrm_spec.rb`) simplify verify that the correct TGT can be acquired from the Samba AD using `kinit` using the domain administrator account `Administrator@BOLT.TEST`.

Reviewer: ```suggestion
Automated tests (in `spec/bolt/transport/winrm_spec.rb`) verify that the correct TGT (ticket granting ticket) can be acquired from the Samba AD using `kinit` using the domain administrator account `Administrator@BOLT.TEST`.
```
